### ISSUE TITLE
LLVM: Improve codegen for Vector<T>.EqualsAll

### ISSUE THREAD

**Issue by EgorBo** [mono/mono#16935]
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static bool Test(Vector<float> a, Vector<float> b) => a == b;
```
Here is what Mono-LLVM JIT emits:
```asm
   0:   c4 e1 f9 6e c6          vmovq  %rsi,%xmm0
   5:   c4 e1 f9 6e cf          vmovq  %rdi,%xmm1
   a:   c5 f1 6c c0             vpunpcklqdq %xmm0,%xmm1,%xmm0
   e:   c4 e1 f9 6e c9          vmovq  %rcx,%xmm1
  13:   c4 e1 f9 6e d2          vmovq  %rdx,%xmm2
  18:   c5 e9 6c c9             vpunpcklqdq %xmm1,%xmm2,%xmm1
  1c:   c5 f8 c2 c1 00          vcmpeqps %xmm1,%xmm0,%xmm0
  21:   c4 e3 79 04 c8 0e       vpermilps $0xe,%xmm0,%xmm1
  27:   c5 f0 54 c0             vandps %xmm0,%xmm1,%xmm0
  2b:   c4 e3 79 04 c8 e5       vpermilps $0xe5,%xmm0,%xmm1
  31:   c5 f8 54 c1             vandps %xmm1,%xmm0,%xmm0
  35:   c4 e3 79 14 c0 00       vpextrb $0x0,%xmm0,%eax
  3b:   c3                      retq
```
The codegen looks a bit odd.
Ideally it should be something like this
```asm
; both vectors are passed via xmm registers
vcmpneqps xmm0, xmm0, xmm1 ; compare two vectors
vpmovmskb eax, xmm0 ; movemask for the result (eax will be 0 if vectors are equal)
test      ax, ax
sete      al
ret
```

LLVM IR we emit:
```llvm
; ModuleID = 'jit-module-Test'
source_filename = "jit-module-Test"

; Function Attrs: noinline uwtable
define monocc i8 @"Program:Test"(i64 %arg_a, i64, i64 %arg_b, i64) #0 {
BB0:
  %2 = alloca <4 x float>, align 16
  %3 = alloca <4 x float>, align 16
  %4 = bitcast <4 x float>* %3 to i8*
  %5 = getelementptr i8, i8* %4, i32 0
  %6 = bitcast i8* %5 to i64*
  store i64 %arg_a, i64* %6
  %7 = getelementptr i8, i8* %4, i32 8
  %8 = bitcast i8* %7 to i64*
  store i64 %0, i64* %8
  %9 = load <4 x float>, <4 x float>* %3
  %10 = bitcast <4 x float>* %2 to i8*
  %11 = getelementptr i8, i8* %10, i32 0
  %12 = bitcast i8* %11 to i64*
  store i64 %arg_b, i64* %12
  %13 = getelementptr i8, i8* %10, i32 8
  %14 = bitcast i8* %13 to i64*
  store i64 %1, i64* %14
  %15 = load <4 x float>, <4 x float>* %2
  br label %BB3

BB3:                                              ; preds = %BB0
  br label %BB2

BB2:                                              ; preds = %BB3
  %16 = fcmp fast oeq <4 x float> %9, %15
  %17 = sext <4 x i1> %16 to <4 x i8>
  %18 = shufflevector <4 x i8> %17, <4 x i8> undef, <4 x i32> <i32 2, i32 3, i32 0, i32 0>
  %19 = and <4 x i8> %17, %18
  %20 = shufflevector <4 x i8> %19, <4 x i8> undef, <4 x i32> <i32 1, i32 0, i32 0, i32 0>
  %21 = and <4 x i8> %19, %20
  %22 = extractelement <4 x i8> %21, i32 0
  br label %BB1

BB1:                                              ; preds = %BB2
  ret i8 %22
}

attributes #0 = { noinline uwtable }
```
We pass the vectors as two `i64` arguments each (struct promotion?) so the func has four i64 arguments. Can we pass them as the real LLVM vectors: `<float x 4>` or `<float x 8>` ?

Godbolt to play: https://godbolt.org/z/e-2O3f (contains reference C++ impl under the one we emit).

**Comment by filipnavara** [comment]
Yeah, I noticed this too... the Mini JIT generates `pmovmskb` ;)

**Comment by filipnavara** [comment]
The code gen that causes it is `OP_XEQUAL` in mini-llvm. The code that you want to generate is the one generated in `OP_EXTRACT_MASK`. The `[v]pmovmskb` instruction exists since the original SSE and it is always present on x64.

**Comment by EgorBo** [comment]
there are two problems here:
1) movmsk should be used
2) ideally, arguments should be passed as `<float x 4>`

**Comment by EgorBo** [comment]
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static Vector<float> Foo1(
    Vector<float> vec1, Vector<float> vec2)
{
    return vec1 + vec2;
}
```
vs
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static Vector<float> Foo2(
    long a, long b, long c, long d, long e, // fake args
    Vector<float> vec1, Vector<float> vec2)
{
    return vec1 + vec2;
}
```

#### Asm for Foo1:
```asm
   0:   c4 e1 f9 6e c2          vmovq  xmm0,rdx
   5:   c4 e1 f9 6e ce          vmovq  xmm1,rsi
   a:   c5 f1 6c c0             vpunpcklqdq xmm0,xmm1,xmm0
   e:   c4 c1 f9 6e c8          vmovq  xmm1,r8
  13:   c4 e1 f9 6e d1          vmovq  xmm2,rcx
  18:   c5 e9 6c c9             vpunpcklqdq xmm1,xmm2,xmm1
  1c:   c5 f8 58 c1             vaddps xmm0,xmm0,xmm1
  20:   c5 f8 11 07             vmovups XMMWORD PTR [rdi],xmm0
  24:   c3                      ret
```

#### Asm for Foo2:
```asm
   0:   c5 f8 28 44 24 08       vmovaps xmm0,XMMWORD PTR [rsp+0x8]
   6:   c5 f8 58 44 24 18       vaddps xmm0,xmm0,XMMWORD PTR [rsp+0x18]
   c:   c5 f8 11 07             vmovups XMMWORD PTR [rdi],xmm0
  10:   c3                      ret
```

For **Foo2** we pass Vectors via stack, so a vector is passed (LLVM IR):
**Foo1**: `i64 %arg_vec1, i64`  ;; two general purpose registers -> copy to xmm
**Foo2**: `<4 x float>* byval %arg_vec1`  ;; stack -> xmm
Ideally it should be passed via xmm by default (`<4 x float>`)

**Comment by SamMonoRT** [comment]
@imhameed - please see if this is fixed by #16984.
